### ISSUE TITLE
401 Error

### ISSUE THREAD

**Issue by thechrisreese** [homespun/homebridge-platform-ring-video-doorbell#16]
I am getting "refresh1 error: Error: API returned status code 401"now constantly.

**Comment by geofbuck** [comment]
Me too constantly and I am not sure if it's related but home bridge not working as well. I deleted the home bridge accessory and now I can't re-add it:  
[5/30/2017, 7:30:44 PM] [Doorbell] refresh1 error: Error: API returned Status Code 401

**Comment by ronzelver** [comment]
Here the same and it results in a crash/restart of homebridge. 

homebridge-0 [Wed May 31 2017 21:47:44 GMT+0000 (UTC)] [Doorbell] _roundTrip error: HTTP response 401
homebridge-0 [Wed May 31 2017 21:47:44 GMT+0000 (UTC)] [Doorbell] refresh2 error: Error: HTTP response 401
homebridge-0 Error: HTTP response 401
homebridge-0     at IncomingMessage.<anonymous> (/usr/local/lib/node_modules/homebridge-platform-ring-video-doorbell/node_modules/homespun-discovery/utilities/utility-roundtrip.js:45:25)
homebridge-0     at emitNone (events.js:85:20)
homebridge-0     at IncomingMessage.emit (events.js:179:7)
homebridge-0     at endReadableNT (_stream_readable.js:913:12)
homebridge-0     at _combinedTickCallback (node.js:377:13)
homebridge-0     at process._tickDomainCallback (node.js:425:11)
PM2 App [homebridge] with id [0] and pid [32776], exited with code [255] via signal [null]
PM2 Starting execution sequence in -fork mode- for app name:homebridge id:0
PM2

**Comment by mrose17** [comment]
i looked at my logs and see the 401. however, my homebridge instance has been running for at week without crashing. 

my guess is that something changed in the ring protocol so that 401 is now being returned.

@davglass - have you noticed anything?

@ronzelver - what version of the plugin are you running?

**Comment by ronzelver** [comment]
I have version 0.2.7 of the plugin installed

**Comment by ronzelver** [comment]
I've just re-installed the plugin and it's back working normal, no more 401 errors in the logs.

**Comment by fserruto** [comment]
Same, reinstall resolved 401 error.

**Comment by kieranmjones** [comment]
I'm having the same problem, a re-install doesn't appear to have fixed the problem though sadly.

**Comment by ronzelver** [comment]
Although homebridge does no longer crash due to the 401 error, notifications from the Ring doorbell are no longer processed.

**Comment by thechrisreese** [comment]
I'm wondering why we are getting unauthorized access. I assume they found out the backdoor to their API in this method and blocked it.

**Comment by thechrisreese** [comment]
Ok with some console logging - this plugin is pinging Ring's API every 4 seconds to check for state changes. I believe this is because the API is unofficial and has no push notifications or way to identify state changes without a "get" request (@homespun ??). 

I have a feeling the 401 unauthorized is coming form our token's being blocked because of calling the API too much. That is why logging out and logging back in by restarting Homebridge will allow it to work for a little while then it stops again.

**Comment by mrose17** [comment]
@thechrisreese - i think your analysis is spot on. if you look at https://github.com/homespun/homebridge-platform-ring-video-doorbell#configuration you'll see there's a `ttl` option. try setting it to something higher than 4. i'm going to try 6 and see what happens. i really do wish there was a better API that was push, rather than pull, driven...

**Comment by thechrisreese** [comment]
@mrose17 any luck with upping ttl?

**Comment by mrose17** [comment]
no. i moved it to 7 seconds and i'm still seeing it. normally, i would expect to see a 429 rather than a 401, so i'm not sure what logic is being used by the folks at ring in order to determine failure. i will now move it to 10 and see what happens there. the odd thing is that using 4 used to work just fine...

**Comment by craigdent** [comment]
any updates on this...? im getting no notifications through homekit anymore whenever motion is detected.. this seems to have happened only recently.. maybe the last week... but have made no changes to cause this..?

**Comment by thechrisreese** [comment]
@craigdent Ring changed something on their server side. We are now being timed out and our auth token is being blocked. So, no you won't get notifications and you will get 401 errors. Still no fix.

**Comment by mruiter** [comment]
any news on the api decode / wiresniff ?

**Comment by thechrisreese** [comment]
I abandoned this and re-added my Ring to my Wink hub and using homebridge-wink3 to access the motion detection.

**Comment by mruiter** [comment]
Users of this API are only reporting video api changes the rest works

https://github.com/davglass/doorbot

**Comment by mrose17** [comment]
@mruiter - i'm having a little difficulty following the situation. is there something i can change in this homebridge plugin in order to make it more reliable?

**Comment by davglass** [comment]
I just pushed up `doorbot@2.0.0` with a fix that works for me now (and a fix for the video api).

The token seems to get invalidated pretty quickly or lost on their side. So I added a retry on the `401` response and I go get a new token. I set my retry to 15 times with a 500ms delay and it seems to make it stable again for me.

The Video API issue wasn't really an API issue, it was a "Javascript sucks at large numbers" issue. The `id` was being rounded during `JSON.parse` since it was too large for JS to handle. I has to cast it to a `string` so that it wasn't messed with.
